broadinstitute/single_cell_portal_core
# Fix cell filtering after clustering switch (SCP-5297)

This makes the cell filtering UX more robust, refactors code for clarity, and adds an automated test.

### Before
Previously, a few bugs occurred when you selected a filter, then went back to the Options panel, then selected another value in the "Clustering" menu.  First, you'd often see an error message `TypeError: Cannot read properties of undefined (reading 'includes')`.  The plot legend would also disappear.  Subsequent cell filtering also showed a permanently stuck "Filtered" legend entry.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/14b19a6a-4cc5-4ba7-b43a-0484d1813ef1

### After
Now, that all works as expected!

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/4b166950-1343-4a3a-a700-680e79d8e448

Another UX refinement: changing the "Color by" annotation no longer clears filters.  Doing so is unnecessary, as cells are the same across annotations.  Preserving filters across color changes allows nimbler exploration.

Behind these UX changes, code quality has also been improved.  Earlier prototype code had generally referred to the "Cell filtering" components as `FacetFilter<Foo>`.  To simplify the cognitive connection between this code and that UI display, the code now refers to the "Cell filtering" components as `CellFiltering<Foo>`, e.g. `CellFilteringPanel`.  This also disambiguates cell filtering code -- which enables faceted search for plotted cells within a study -- to the "Advanced search" code -- which also sometimes uses the term `FacetFilter` in code which enables faceted search for studies across the entire application.

### Test
A new automated test verifies basic initialization of the cell filtering panel React component.  To manually test:
* If on staging, go to [SCP303](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303); if local, go to "Human milk - differential expression" DE pilot
* Enable cell filtering flag if it's not already
* Click "Cell filtering", select a filter, then deselect it, then click the "<-" button to go back to the Options panel
* Select a different clustering in the "Clustering" menu
* Confirm you don't see an error message "TypeError: Cannot read properties of undefined (reading 'includes')`.", and that  the plot legend would appears.

This relates to SCP-5297.
